code/analyticsedge/src/phone/java/com/adobe/marketing/mobile/AnalyticsExtension.java

```diff
@@ -234,47 +316,53 @@ private String getActionPrefix(final boolean isInternalAction) {
+                AnalyticsConstants.ContextDataKeys.ACTION;
     }
 
     /**
      * This prepares the analytics variables and analytics data from the passed in event.
      *
-     * @param event The Generic Track Request Content {@link Event}.
+     * @param event An event containing track data for processing
+     * @param data Track data for processing
      */
-    private void track(final Event event) {
-        final HashMap<String, String> analyticsVars = processAnalyticsVars(event);
-        final HashMap<String, String> analyticsData = processAnalyticsData(event);
+    private void track(final Event event, final Map<String, Object> data) {
+        if(!(data.containsKey(AnalyticsConstants.EventDataKeys.TRACK_STATE) ||
+                data.containsKey(AnalyticsConstants.EventDataKeys.TRACK_ACTION) ||
+                data.containsKey(AnalyticsConstants.EventDataKeys.CONTEXT_DATA))) {
+            Log.warning(LOG_TAG, "track - Dropping request as event data is missing state, action or contextData");
+            return;
+        }
+        final HashMap<String, String> analyticsVars = processAnalyticsVars(event, data);
+        final HashMap<String, String> analyticsData = processAnalyticsData(event, data);
         sendAnalyticsHit(analyticsVars, analyticsData);
     }
 
     /**
      * This method converts the event's event data into analytics variables.
      *
      * @param event The Generic Track Request Content {@link Event}.
+     * @param data Track data for processing
      *
      * @return {@code Map<String, String>} containing the vars data
      */
-    private HashMap<String, String> processAnalyticsVars(final Event event) {
+    private HashMap<String, String> processAnalyticsVars(final Event event, final Map<String, Object> data) {
         final HashMap<String, String> processedVars = new HashMap<>();
-        final EventData eventData = event.getData();
-
-        if(eventData == null || eventData.isEmpty()) {
-            return processedVars;
-        }
-
-        // context: pe/pev2 values should always be present in track calls if there's action regardless of state.
+        // Context: pe/pev2 values should always be present in track calls if there's action regardless of state.
         // If state is present then pageName = state name else pageName = app id to prevent hit from being discarded.
-        final String actionName = eventData.optString(AnalyticsConstants.EventDataKeys.TRACK_ACTION, null);
+        final String actionName = (String) data.get(AnalyticsConstants.EventDataKeys.TRACK_ACTION);
         if(!StringUtils.isNullOrEmpty(actionName)) {
             processedVars.put(AnalyticsConstants.AnalyticsRequestKeys.IGNORE_PAGE_NAME, AnalyticsConstants.IGNORE_PAGE_NAME_VALUE);
-            boolean isInternal = eventData.optBoolean(AnalyticsConstants.EventDataKeys.TRACK_INTERNAL, false);
+            boolean isInternal = false;
```

Comment:
Any reason for not using opt?

Reply:
`data` is a Map in this method.

Reply:
In that case can we use map.getOrDefault(), it'll be much cleaner.

Reply:
ah, tried this and getOrDefault() requires an API level of 24. we currently have a min of 14.